src/Storages/MergeTree/ZooKeeperRetries.h

```diff
@@ -0,0 +1,250 @@
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int OK;
+}
+
+struct ZooKeeperRetriesInfo
+{
+    ZooKeeperRetriesInfo() = default;
+    ZooKeeperRetriesInfo(std::string name_, Poco::Logger * logger_, UInt64 max_retries_, UInt64 initial_backoff_ms_, UInt64 max_backoff_ms_)
+        : name(std::move(name_))
+        , logger(logger_)
+        , max_retries(max_retries_)
+        , curr_backoff_ms(std::min(initial_backoff_ms_, max_backoff_ms_))
+        , max_backoff_ms(max_backoff_ms_)
+    {
+    }
+
+    std::string name;
+    Poco::Logger * logger = nullptr;
+    UInt64 max_retries = 0;
+    UInt64 curr_backoff_ms = 0;
+    UInt64 max_backoff_ms = 0;
+    UInt64 retry_count = 0;
+};
+
+class ZooKeeperRetriesControl
+{
+public:
+    ZooKeeperRetriesControl(std::string name_, ZooKeeperRetriesInfo & retries_info_) : name(std::move(name_)), retries_info(retries_info_)
+    {
+    }
+
+    bool canTry()
+    {
+        ++iteration_count;
+        /// first iteration is ordinary execution, no further checks needed
+        if (0 == iteration_count)
+            return true;
+
+        if (unconditional_retry)
+        {
+            unconditional_retry = false;
+            return true;
+        }
+
+        /// iteration succeeded -> no need to retry
+        if (iteration_succeeded)
+        {
+            if (retries_info.logger)
+                LOG_DEBUG(
+                    retries_info.logger,
+                    "ZooKeeperRetriesControl: {}/{}: succeeded after: iterations={} total_retries={}",
+                    retries_info.name,
+                    name,
+                    iteration_count,
+                    retries_info.retry_count);
+            return false;
+        }
+
+        if (stop_retries)
+        {
+            logLastError("stop retries on request");
+            action_after_last_failed_retry();
+            throwIfError();
+            return false;
+        }
+
+        if (retries_info.retry_count >= retries_info.max_retries)
+        {
+            logLastError("retry limit is reached");
+            action_after_last_failed_retry();
+            throwIfError();
+            return false;
+        }
+
+        /// retries
+        ++retries_info.retry_count;
+        logLastError("will retry due to error");
+        sleepForMilliseconds(retries_info.curr_backoff_ms);
+        retries_info.curr_backoff_ms = std::min(retries_info.curr_backoff_ms * 2, retries_info.max_backoff_ms);
+
+        /// reset the flag, it will be set to false in case of error
+        iteration_succeeded = true;
+
+        return true;
+    }
+
+    void retryLoop(auto && f)
+    {
+        while (canTry())
+        {
+            try
+            {
+                f();
+            }
+            catch (const zkutil::KeeperException & e)
+            {
+                if (!Coordination::isHardwareError(e.code))
+                    throw;
+
+                setKeeperError(e.code, e.message());
+            }
```

Comment:
It will not retry on `TABLE_IS_READ_ONLY` (`StorageReplicatedMergeTree::write` may throw this)

Reply:
From `StorageReplicatedMergeTree::write`